src/algorithms/walk.cpp

```diff
@@ -130,6 +130,36 @@ uint64_t walk_haplotype_frequency(const HandleGraph& graph,
     return search_state.size();
 }
 
+std::vector<std::string> walk_haplotype_names(const HandleGraph& graph,
+                                              const gbwt::GBWT& haplotypes,
+                                              const walk_t& walk) {
+    std::vector<std::string> names;
+    if (walk.path.empty()) {
+        return names;
+    }
+    auto& first_step = walk.path.front();
+    gbwt::node_type start_node = gbwt::Node::encode(graph.get_id(first_step), graph.get_is_reverse(first_step));
+    gbwt::SearchState search_state = haplotypes.find(start_node);
+    for (uint64_t i = 1; i < walk.path.size(); ++i) {
+        auto& next = walk.path[i];
+        gbwt::node_type next_node = gbwt::Node::encode(graph.get_id(next), graph.get_is_reverse(next));
+        search_state = haplotypes.extend(search_state, next_node);
+        if (search_state.empty()) {
+            break;
+        }
+    }
+    assert(haplotypes.hasMetadata() && haplotypes.metadata.hasSampleNames());
+    for (auto& thread : haplotypes.locate(search_state)) {
+        // not clear why we have to do this, but it's necessary
+        // perhaps due to fwd/reverse threads
+        auto id = (thread-1)/2; // convert thread id returned by locat to gbwt metadata compatible one
```

Comment:
This should be `auto id = thread / 2` or `auto id = gbwt::Path::id(thread)`. The GBWT sequence ids returned by `locate()` encode VG thread id (or GBWT path id) and orientation much in the same way as GBWT node ids encode VG node id and orientation. It would not make much sense to store duplicate path names in a bidirectional GBWT, and it would violate the principle that path names should be unique.

Reply:
Thanks @jltsiren !